src/peer_select.cc

```diff
@@ -975,53 +1237,27 @@ PeerSelector::url() const
     return noUrl;
 }
 
-/// \returns valid/interested peer initiator or nil
-PeerSelectionInitiator *
-PeerSelector::interestedInitiator()
-{
-    const auto initiator = initiator_.valid();
-
-    if (!initiator) {
-        debugs(44, 3, id << " initiator gone");
-        return nullptr;
-    }
-
-    if (!initiator->subscribed) {
-        debugs(44, 3, id << " initiator lost interest");
-        return nullptr;
-    }
-
-    debugs(44, 7, id);
-    return initiator;
-}
-
-bool
-PeerSelector::wantsMoreDestinations() const {
-    const auto maxCount = Config.forward_max_tries;
-    return maxCount >= 0 && foundPaths <
-           static_cast<std::make_unsigned<decltype(maxCount)>::type>(maxCount);
-}
-
 void
-PeerSelector::handlePath(Comm::ConnectionPointer &path, FwdServer &fs)
+PeerSelector::callback(CachePeer *peer, const hier_code code)
 {
-    ++foundPaths;
+    request->hier.ping = ping;
 
-    path->peerType = fs.code;
-    path->setPeer(fs._peer.get());
-
-    // check for a configured outgoing address for this destination...
-    getOutgoingAddress(request, path);
-
-    request->hier.ping = ping; // may be updated later
-
-    debugs(44, 2, id << " found " << path << ", destination #" << foundPaths << " for " << url());
+    if (code == HIER_NONE)
+        debugs(44, 2, id << " found all " << foundPeers << " peers for " << url());
+    else
+        debugs(44, 2, id << " found peer " << (peer ? peer->name : "direct") << ", " << foundPeers << " peer for " << url());
```

Comment:
I believe it is better to not fill the code with ifs just for debug messages. 
The messages as is are enough to make someone understand what is happens here. The only we should change is the debug level. However because the following (untouched) debugs are  in level 2 I left these debugs to level 2.